stmtbundle: refactor FK handling in schema.sql

yuzefovich opened the pull request:
This PR refactors how we include tables and FKs into the bundles. It also optimizes _which_ tables we include. See each commit for details.

Fixes: #139196.
Fixes: #142006.

yuzefovich committed:
sql: add IGNORE_FOREIGN_KEYS option to SHOW CREATE TABLE

This will be used in the following commit for stmt bundle construction.

`SHOW CREATE TABLE` command has been extended with `WITH IGNORE_FOREIGN_KEYS`
option which makes it so that we use existing `create_nofks` column of
`crdb_internal.create_statements` virtual table. This new option cannot
be combined with existing `WITH REDACT` option (because we don't have
the corresponding column, at least yet).

Release note (sql change): New `WITH IGNORE_FOREIGN_KEYS` option is
added to `SHOW CREATE TABLE` stmt which makes it so that the foreign key
constraints are not included in the output schema. This option is also
acceptable in `SHOW CREATE VIEW` but has no influence there. It cannot
be combined with existing `WITH REDACT` option.

yuzefovich committed:
stmtbundle: refactor FK handling in schema.sql

As of 982abc239efa69a808e4a2b4ecbb01126b253d8f, we include all FK
referenced and referencing tables, including transitively, into the stmt
bundles for all queries. This information might be required to
understand the plans in some cases (like when we have a mutation which
results in a CASCADE post-query), so we initially went with "include
everything all the time". This can be suboptimal since we could pull in
information that isn't actually needed for a particular query.
Additionally, we've just seen a case where in the multi-region cluster
pulling in all these related tables was taking minutes.

To partially alleviate this problem, this commit refactors how we
include FKs into the bundles. Namely, unless a redacted bundle is
requested, we'll use newly-added `WITH IGNORE_FOREIGN_KEYS` option of
`SHOW CREATE TABLE`; then we'll add only relevant FKs as ALTER TABLE
statements. For read-only queries we now include directly referenced
tables as well as FKs between them (i.e. both origin and referenced
tables of a FK constraint must be directly referenced by the metadata to
be included); for mutations we still include all referenced and
referencing tables, including transitively, as well as all FKs. Reducing
the set of tables for mutations is left as a follow-up.

As a bonus, this refactor allows us to gracefully handle FK cycles (i.e.
the bundles will be recreatable).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142018)
<!-- Reviewable:end -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice improvement!

> Reviewed 6 of 6 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-:-OK32oz4AXwl62Y5dIy5:b-6f8vr5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/explain_bundle.go` line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK323gaEqb_uGuRhX_c:-OK323gaEqb_uGuRhX_d:bnf27lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/explain_bundle.go#L588)):*
> ```Go
> 			ctx,
> 			b.plan.catalog,
> 			func(table cat.Table, directlyReferenced intsets.Fast) bool {
> ```

nit: I think it'd be nicer to add a method to get the directly referenced tables (or alternatively, use `NumTables()` and `Table()` to get the list)


<!-- Sent from Reviewable.io -->

yuzefovich committed:
stmtbundle: optimize FK inclusion for mutations

This commit further optimizes which FK reference tables we include into
the stmt bundle for mutation statements. We always include all tables
that are directly referenced by the metadata. Then for each foreign key
we need to decide on two actions:
1) whether to include the reference table, and
2) whether to "recurse" into the reference table (i.e. whether FKs of
the reference table should be "explored").

This commit implements the following logic:
- if we're considering a referenced table of a FK. This means that the
origin table has already been included (i.e. we included `child` and are
deciding on whether to include `parent`). Only for DELETEs we don't want
to include the `parent` because it is unaffected, for all other
mutations we do need to perform the FK check against `parent` when
modifying `child`. (Note that strictly speaking we only need to include
`parent` if the FK columns in `child` are actually modified by the stmt,
but it seems tricky to get that right, so we don't try.)
- if we're considering an origin table of a FK. This means that the
referenced table has already been included (i.e. we included `parent`
and are deciding on whether to include `child`).
  - if we have an INSERT into `parent`, then we don't need to include
`child`
  - if we have DELETE, UPDATE, or UPSERT, then we always will need to
 perform the FK check, so we always include `child`.
    - additionally, if we have a relevant CASCADE action, we also want
to "recurse" into `child` since we might execute a CASCADE postquery
which could trigger more FK checks / cascades.

Note that some of the FK reference tables are already directly
referenced by the metadata, so we might end up including an extra
"layer" of FK references. Avoiding that seems quite tricky, so we don't
do that.

This commit additionally adds an escape hatch - a cluster setting that
allows us to include all FK references unconditionally.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142018)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/explain_bundle.go` line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK323gaEqb_uGuRhX_c:-OK5NEHF1tKHjHfufmfC:b-t665) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/explain_bundle.go#L588)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: I think it'd be nicer to add a method to get the directly referenced tables (or alternatively, use `NumTables()` and `Table()` to get the list)
</blockquote></details>

I wrote it this way since I was already thinking about how to improve the situation for mutations. I just pushed another commit that does that - PTAL and let me know whether this approach makes sense.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Just added another commit that I will squash into the third. I kept it separate for now since I think it'll make reviewing the logic easier.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-:-OK7d2rrAx4zFOvhuFVv:b50mua1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 15 of 15 files at r2, 10 of 10 files at r4, 3 of 11 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142018)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @michae2, and @yuzefovich)

___
*[`pkg/sql/explain_bundle.go` line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK323gaEqb_uGuRhX_c:-OK8J_hsCjI_AxfGWOA9:b-qiqcii) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/explain_bundle.go#L588)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I wrote it this way since I was already thinking about how to improve the situation for mutations. I just pushed another commit that does that - PTAL and let me know whether this approach makes sense.
</blockquote></details>

I think this part of the metadata API has grown unruly. I'd prefer the metadata API was simpler, e.g., it simply could tell you things about the objects in the query plan, and users of the API could filter those and do what they'd like with the information. Passing callbacks with multiple parameters makes this stmt bundle logic and metadata more tightly coupled than it needs to be. 

I think the metadata could also provide the list of tables (maybe an int set of ordinals into `md.tables`) that are inserted, updated, deleted, upserted, to make traversing the FK tree easier, and making the plan flags added in this PR unnecessary.

That being said, this coupling didn't start with this PR, so it doesn't need to be the one to fix it.
___
*[`pkg/sql/opt/metadata.go` line 1120 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK8FrCN2pDCd95i6r57:-OK8FrCN2pDCd95i6r58:b278n7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3a5bcd8f9c95c0877ddf5e774f56a8836c0bf8/pkg/sql/opt/metadata.go#L1120)):*
> ```Go
> 	var directlyReferenced intsets.Fast
> 	for _, table := range md.tables {
> 		directlyReferenced.Add(int(table.Table.ID()))
> ```

I'm confused how this works. `md.tables` should contain FK tables that are not directly referenced.

...I think it works for the wrong reasons: `md.tables` includes FK tables that require a post-query check, which we want to always include. And it doesn't contain FK tables that we build post-query cascades, which we only want to include sometimes.

~In that case, would it be simpler to solve this by adding FK tables that we _will_ build cascades for to the metadata (either `md.tables` or another field) and just use that to get all tables for the bundle. It seems like it should work to me—if the query plan didn't build anything for an FK table, then it doesn't need to be in the bundle.~ Nvm, this won't work for the recursive cases.

I think we should at least rename this variable/return value to make it more clear that these tables may not actually be directly referenced.
___
*[`pkg/sql/explain_bundle_test.go` line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK89Dx78QSctroyPrbg:-OK89Dx78QSctroyPrbh:brupek3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/explain_bundle_test.go#L382)):*
> ```Go
> 		// Now read different combinations of tables which will influence
> 		// whether ADD CONSTRAINT ... FOREIGN KEY statements are included.
> 		targetTableNames = []string{"parent", "child1"}
> ```

nit: change `contentCheck` into something like `contentCheckFn(targetTableNames, targetFKs []string) (func(name, contents string) error) { ... }`. Modifying these shared variables is hard to follow and error prone, IMO.
___
*[`pkg/sql/plan.go` line 658 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK8Ctu-3AvfEzMU4Cwa:-OK8Ctu-3AvfEzMU4Cwb:bqycieo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3a5bcd8f9c95c0877ddf5e774f56a8836c0bf8/pkg/sql/plan.go#L658)):*
> ```Go
> 
> 	// planFlagContainsMutation is set if the plan has any mutations.
> 	planFlagContainsMutation
> ```

nit: you can remove this and add `res.flags.Set(planFlagContainsMutation)` to each of the 4 mutation branches in `constructPlan`.
___
*[`pkg/sql/logictest/testdata/logic_test/show_create` line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK86Hox2-Vw0NDjqg1y:-OK86Hox2-Vw0NDjqg1z:b-a4v5gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/logictest/testdata/logic_test/show_create#L146)):*
> ```
>      CONSTRAINT unique_a_partial UNIQUE WITHOUT INDEX (a) WHERE b > 0:::INT8,
>      CONSTRAINT unique_b UNIQUE WITHOUT INDEX (b),
>      CONSTRAINT unique_b_partial UNIQUE WITHOUT INDEX (b) WHERE a > 0:::INT8
> ```

nit: add a foreign key to this table to prove that it is omitted


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142018)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/explain_bundle.go` line 588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK323gaEqb_uGuRhX_c:-OKDWUjJ0vhWRCez3K4E:b-mmdyyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/explain_bundle.go#L588)):*
I agree that it has gotten too tightly coupled (mostly I'm the one to blame for this). I did a refactor to pull out a visitor method (although it still requires two callbacks), but it's no longer connected to the `opt.Metadata` struct. I just pushed it at #142151 and the refactor is present in the second commit there.

Please take a look and let me know what you think. I'm leaning on the side of proceed with #142151 and backporing that as well.

> I think the metadata could also provide the list of tables (maybe an int set of ordinals into `md.tables`) that are inserted, updated, deleted, upserted, to make traversing the FK tree easier, and making the plan flags added in this PR unnecessary.

I agree that this would be nice, but it's outside of my comfort zone and seems more involved and error-prone, so I didn't look into this and only added a TODO.
___
*[`pkg/sql/plan.go` line 658 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK8Ctu-3AvfEzMU4Cwa:-OKDT-6XCaPEhQ_3C1Sj:byy3tjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3a5bcd8f9c95c0877ddf5e774f56a8836c0bf8/pkg/sql/plan.go#L658)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: you can remove this and add `res.flags.Set(planFlagContainsMutation)` to each of the 4 mutation branches in `constructPlan`.
</blockquote></details>

Hm, I think `planFlagContainsMutation` in the current form is super set of `DELETE || INSERT || UPDATE || UPSERT` - mutations plan flag is also set for different ALTER and CREATE statements. I added a TODO to look into this.
___
*[`pkg/sql/opt/metadata.go` line 1120 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK8FrCN2pDCd95i6r57:-OK8mSL_-I6GYqXb7tut:b-dmw4tm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3a5bcd8f9c95c0877ddf5e774f56a8836c0bf8/pkg/sql/opt/metadata.go#L1120)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm confused how this works. `md.tables` should contain FK tables that are not directly referenced.

...I think it works for the wrong reasons: `md.tables` includes FK tables that require a post-query check, which we want to always include. And it doesn't contain FK tables that we build post-query cascades, which we only want to include sometimes.

~In that case, would it be simpler to solve this by adding FK tables that we _will_ build cascades for to the metadata (either `md.tables` or another field) and just use that to get all tables for the bundle. It seems like it should work to me—if the query plan didn't build anything for an FK table, then it doesn't need to be in the bundle.~ Nvm, this won't work for the recursive cases.

I think we should at least rename this variable/return value to make it more clear that these tables may not actually be directly referenced.
</blockquote></details>

I renamed the variable. Indeed, I was initially unaware that tables that we need to do FK checks for are referenced by the metadata.
___
*[`pkg/sql/logictest/testdata/logic_test/show_create` line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK86Hox2-Vw0NDjqg1y:-OK8j4iWBHELWMmEqo6C:bhuf9qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/logictest/testdata/logic_test/show_create#L146)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a foreign key to this table to prove that it is omitted
</blockquote></details>

There is already `fk_a` that referenced `d(d)` (it is shown in `SHOW CREATE TABLE c WITH REDACT` in the test case right above this newly added one).
___
*[`pkg/sql/explain_bundle_test.go` line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142018#-OK89Dx78QSctroyPrbg:-OK8kwg5-SKCeQ-TDDe1:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd44dd249cd8921557ba110bbe121957f9d963e5/pkg/sql/explain_bundle_test.go#L382)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: change `contentCheck` into something like `contentCheckFn(targetTableNames, targetFKs []string) (func(name, contents string) error) { ... }`. Modifying these shared variables is hard to follow and error prone, IMO.
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
superseded by #142151
